README.md
# TweetPull
Create an RSS feed of a selected Twitter user's recent tweets using the v1.1 API.

## Background
For some time, Twitter RSS-enabled a user's stream, meaning one could essentially follow a user without interacting with the Twitter front-end website (or app). This allowed one to add one of two specifically-crafted URLs into their RSS reader of choice:

`http://twitter.com/statuses/user_timeline/[numeric twitter id].rss` , or 
`http://api.twitter.com/1/statuses/user_timeline.rss?screen_name=[twitter screen name]`

This was great for people like me who are devoted to RSS, want to follow people on Twitter, but generally hate sifting through the streaming mess of the Twitter site. However, in the summer of 2013, Twitter officially deprecated v1 of its API, and released [v1.1](https://dev.twitter.com/rest/public); and with it, they removed the ability to make un-authenticated requests to the API. This effectively killed any ability to quickly and easily pull recent tweets via a simple URL.

## What You'll Need
- PHP
- a web host for the RSS feeds
- a scheduler (like cron) to run the PHP script

## My Approach
Since the v1.1 API allows authenticated requests to a large number of very useful methods, the first step was to sign in to the [Twitter Developer](https://dev.twitter.com/) site and create an application that would authenticate back to me. This is done via creation of "consumer keys" and "access tokens" that are unique to the developer and are placed in the request header.

The next step was to find a way to create a valid request header with my keys/credentials along with my desired get method (in my case, [statuses/user_timeline](https://dev.twitter.com/docs/api/1.1/get/statuses/user_timeline)). GitHub user **themattharris** made all the authentication portion incredibly easy by developing an [OAuth library written in PHP](https://github.com/themattharris/tmhOAuth).

Once the request was formed, my code was then a matter of extracting the JSON response, parsing the user's tweets, and dropping them to a well-formed RSS/XML file. This might seem like an involved and somewhat unnecessary approach, but my primary reasoning was due to Twitter's [API rate limiting](https://dev.twitter.com/docs/rate-limiting/1.1). Read that, and the following "gotchas" should make more sense:

- Writing to file instead of STDOUT with RSS/XML headers keeps me in the control of the requests, and not busting the limit. I run the script via cron at a [safely] scheduled interval.
- Once a file is written, I need to host it so an RSS reader can get to it. No web host, no usefulness.
- Although there are different ways I could do this, I chose to pass the user id to the script as an argument. This means if I want to follow/read 5 Twitter users, I have 5 entries in my crontab. (See Installation steps below.)

In general, a lot of improvements can be made here, but it's good enough for a v0.1 release. And I only did it because the v1.1 API is so new, I couldn't find anyone else doing something similar yet.

## Installation
### General
1. Download [tmhOAuth library](https://github.com/themattharris/tmhOAuth)
2. Download latest version of TweetPull for RSS
3. Open `tweetpull-config.php` config file and set the following **required** parameters (descriptions for each appear in config file):
    - `OAUTH_PATH`
    - `FILE_PATH`
    - `CONSUMER_KEY, CONSUMER_SECRET, ACCESS_KEY, ACCESS_SECRET`
4. Review and set the optional parameters in the `tweetpull-config.php` config file

### Usage
1. Follow Installation steps above
2. Run the script (typically by cron) with the following:
    ```
    5 * * * * php /full/path/to/tweetpull.php -u [twitter screen name] [-r] [-t]
    ```
    \-r is an optional parameter. If it is set, replies are removed from the returned timeline. See [statuses/user_timeline](https://dev.twitter.com/docs/api/1.1/get/statuses/user_timeline) for more info.
    
    \-t is an optional parameter. If it is set, retweets are removed from the returned timeline. See [statuses/user_timeline](https://dev.twitter.com/docs/api/1.1/get/statuses/user_timeline) for more info.
3. Output file(s) will be \[twitter screen name\].xml. Enter that file (with full URL) into your favorite RSS reader. (e.g. `http://www.mydomain.com/feeds/[twitter screen name].xml`)


CHANGELOG.md
# Changelog
0.4
- added exclude_replies and include_rts (from statuses/user_timeline) to optionally remove replies or retweets

0.3.1
- added generator tag to RSS header output

0.3
- created separate config class to define all params; no modifications to main tweetpull.php script are needed
    - made many features configurable
- various debug levels: info messages, error messages; various debug output options: screen, log file, both
- ability to pull data via sample data response file for testing, instead of live pull from Twitter API
- better comments

0.2.1
- added media entity support in links() function; translates embedded images into HTML IMG tags within RSS content

0.2
- added function to link embedded entities (hyperlinks to site, @mentions to Twitter user, #hashtags to hashtag search)
- better comments
- enforce 15-char max username length, per Twitter policy

0.1
- First released version.

# To-do List
- ~~make embedded links linkable~~
- ~~link @username mentions back to user~~
- ~~link #hastags to tag search~~
- ~~better code comments~~
- ~~link embedded "media" entities~~
- ~~separate config from program logic~~
    - ~~add flags for debugging, enhanced features, errors, etc.~~
- ~~option to turn off retweets; on twitter or via api?~~
- maintain embedded media link/url in addition to IMG
- why are RTs getting truncated? is that me or twitter?


tweetpull-config.php
<?php
/**
 * TweetPull for RSS Config
 * https://github.com/sahearn/TweetPull
 * http://scott.teamahearn.com/tweetpull4rss/
 *
 * @author sahearn
 */

class Config {
    
    /*
     * ************************************************************************
     *  REQUIRED PARAMETERS
     *  you must set these PATH, CONSUMER_, and ACCESS_ parameters
     * ************************************************************************
     */
    
    /* full path to tmhOAuth.php */
    const OAUTH_PATH = '/full/path/to/tmhOAuth/';
    
    /* full path to output RSS file */
    const FILE_PATH = '/full/path/to/output/destination/';
    
    /* keys and tokens for your dev.twitter app */
    const CONSUMER_KEY      = '[your consumer key]';
    const CONSUMER_SECRET   = '[your consumer secret]';
    const ACCESS_TOKEN      = '[your access token]';
    const ACCESS_SECRET     = '[your access secret]';

    
    
    
    /*
     * ************************************************************************
     *  OPTIONAL PARAMETERS
     *  set these next few parameters to your liking
     * ************************************************************************
     */

    /*
     * how to render entities (links, @mentions, #hastags, media) embedded in tweets
     * e.g. http://t.co/abcde becomes <a href="http://t.co/abcde">http://t.co/abcde</a>
     *      @twitterid becomes <a href="https://twitter.com/twitterid">@twitterid</a>
     *      #api becomes <a href="https://twitter.com/search?q=%23api&src=hash">#api</a>
     * 
     * this will only take effect in the RSS <item> description
     * 
     * also see EMBED_IMAGE below for rendering of embedded media/images
     * 
     * 0 => do not convert entities to hyperlinked HTML
     * 1 => convert entities to hyperlinked HTML
     */
    const CONVERT_ENTITIES = 1;
    
    /* 
     * how to render images embedded in tweets
     * e.g. http://t.co/abcde becomes <img src="http://full.path.com/abcde.jpg" width="[width]" height="[height]" border="0">
     * 
     * 0 => do not render images inline in RSS post content; link only
     * 1 => render images inline in RSS post content with HTML IMG tag
     */
    const EMBED_IMAGE = 1;

    /* 
     * for the purposes of RSS, the tweet itself will be used as its own title and 
     * description.  if you'd like the title truncated (safely by word-boundary, 
     * appended with ellipses) to a certain length, set TRUNCATE_TITLE to a non-zero 
     * integer.  set to 0 to leave the title unmodified
     * e.g. 50
     */
    const TRUNCATE_TITLE = 0;

    
    
    
    /*
     * ************************************************************************
     *  DEBUG PARAMETERS
     *  probably no need to mess with anything else down here unless you're
     *  having problems and need to debug
     * ************************************************************************
     */
    
    /* 
     * OFF      => no info or errors to output
     * INFO     => info/messages to DEBUG_OUTPUT
     * ERROR    => errors to DEBUG_OUTPUT
     * ALL      => INFO and ERRORS to DEBUG_OUTPUT
     */
    const DEBUG_MODE = 'OFF';
    
    /*
     * SCREEN   => output to screen
     * LOG      => output to logfile "[twitter user id].log"
     * ALL      => output to both screen and logfile
     * 
     * if DEBUG_MODE = 'OFF', DEBUG_OUTPUT will have no effect
     */
    const DEBUG_OUTPUT = 'SCREEN';
    
    /*
     * LIVE     => production; live call to Twitter API
     * SAMPLE   => sample response; no live call to Twitter API; good for testing
     * 
     * if set to 'SAMPLE', -u command line argument is still required, but
     * will have no effect on retrieved data.  script will use the contents 
     * of SAMPLE_DATA instead.
     */
    const DATA_SOURCE = 'LIVE';
    
    /* location of sample data if DATA_SOURCE is set to 'SAMPLE' */
    const SAMPLE_DATA = "tweetpull-sample-response-success_serialized.txt";
}

?>


tweetpull.php
<?php
/**
 * TweetPull for RSS
 * Create an RSS feed of a selected Twitter user's recent tweets using the v1.1 API.
 * https://github.com/sahearn/TweetPull
 * http://scott.teamahearn.com/tweetpull4rss/
 *
 * @author sahearn
 * @version 0.4
 *
 * 10 April 2014
 *
 *  
 *    Copyright 2014 Scott A'Hearn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 */

$options = getopt("u:r::t::"); 
if (empty($options) ) { 
    print "User id is required. [-u userID]" . "\n\n";
    print "usage: tweetpull.php -u [userID] [-r] [-t]" . "\n\n";
    print "  optional:" . "\n";
    print "\t" . "-r" . "\t" . "exclude replies" . "\n";
    print "\t" . "-t" . "\t" . "exclude retweets" . "\n";
    exit(1); 
}
// clean up input
$passedID = substr(strip_tags(stripslashes($options['u'])), 0, 15);
$passedReplies = false;
if (isset($options['r']) && is_bool($options['r']))
    $passedReplies = true;
$passedRetweets = true;
if (isset($options['t']) && is_bool($options['t']))
    $passedRetweets = false;

// import and initialize config file
include 'tweetpull-config.php';
initConfig();

// import auth library to credential with twitter
require Config::OAUTH_PATH . 'tmhOAuth.php';

class tmhOAuthExample extends tmhOAuth {
    const VERSION = '0.4';

    public $twitter_userid;
    public $twitter_name;
    public $twitter_desc;
    public $twitter_icon;
    public $file;
    
    public function __construct($config = array()) {
        $this->config = array_merge(
            array(
                // change the values below to ones for your application
                'consumer_key'    => Config::CONSUMER_KEY,
                'consumer_secret' => Config::CONSUMER_SECRET,
                'token'           => Config::ACCESS_TOKEN,
                'secret'          => Config::ACCESS_SECRET,
                //'bearer'          => 'YOUR_OAUTH2_TOKEN',
                'user_agent'      => 'TweetPull for RSS ' . self::VERSION . ' - //github.com/sahearn/TweetPull with tmhOAuth ' . parent::VERSION . ' - //github.com/themattharris/tmhOAuth',
            ),
            $config
        );

        parent::__construct($this->config);
    }
  
    // open output file for writing; config with your path here
    public function openOutputFile() {
        $this->file = Config::FILE_PATH . $this->twitter_userid . '.xml';
        writeDebugMessage("INFO", "Output RSS file created: " . $this->file);
    }
  
    public function writeHeader() {
        $output = '<?xml version="1.0" encoding="UTF-8"?>' . "\n";
        $output .= '<rss version="2.0"' . "\n";
        $output .= ' xmlns:content="http://purl.org/rss/1.0/modules/content/"' . "\n";
        $output .= ' xmlns:dc="http://purl.org/dc/elements/1.1/"' . "\n";
        $output .= ' >' . "\n";

        $output .= "\t" . '<channel>' . "\n";
        $output .= "\t\t" . '<title>' . $this->twitter_name . ' (' . $this->twitter_userid . ') on Twitter</title>' . "\n";
        $output .= "\t\t" . '<link>https://twitter.com/' . $this->twitter_userid . '</link>' . "\n";
        $output .= "\t\t" . '<description>' . $this->twitter_desc . '</description>' . "\n";
        $output .= "\t\t" . '<generator>TweetPull for RSS v' . self::VERSION . ' - //bit.ly/15dLYuO</generator>' . "\n";
        $output .= "\t\t" . '<language>en-US</language>' . "\n";

        $output .= "\t\t" . '<image>' . "\n";
        $output .= "\t\t" . ' <title>' . $this->twitter_name . ' (' . $this->twitter_userid . ') on Twitter</title>' . "\n";
        $output .= "\t\t" . ' <url>' . $this->twitter_icon . '</url>' . "\n";
        $output .= "\t\t" . ' <link>https://twitter.com/' . $this->twitter_userid . '</link>' . "\n";
        $output .= "\t\t" . '</image>' . "\n";

        file_put_contents($this->file, $output);
        
        writeDebugMessage("INFO", "Output RSS header written.");
    }
  
    public function writeFooter() {
        $output = "\t" . '</channel>' . "\n";
        $output .= "</rss>";

        $current = file_get_contents($this->file);
        $current .= $output;
        file_put_contents($this->file, $current);
        
        writeDebugMessage("INFO", "Output RSS footer written.");
    }
  
    public function truncate($string, $length, $stopanywhere=false) {
        //truncates a string to a certain char length, stopping on a word if not specified otherwise.
        if (strlen($string) > $length) {
            //limit hit!
            $string = substr($string,0,($length -3));
            if ($stopanywhere) {
                //stop anywhere
                $string .= '...';
            } else{
                //stop on a word.
                $string = substr($string,0,strrpos($string,' ')).'...';
            }
        }
        return $string;
    }

    public function links($r) {
        $final = '';
        $respText = $r["text"];

        $urls = $r["entities"]["urls"];
        foreach ($urls as $url) {
            $uUrl = $url["url"];
            list($uStart, $uEnd) = $url["indices"];
            $entities[] = array( $uStart, $uEnd, 'url', $uUrl );
        }
        
        $mentions = $r["entities"]["user_mentions"];
        foreach ($mentions as $mention) {
            $mName = $mention["screen_name"];
            list($mStart, $mEnd) = $mention["indices"];
            $entities[] = array( $mStart, $mEnd, 'mention', '@' . $mName );
        }
        
        $hashtags = $r["entities"]["hashtags"];
        foreach ($hashtags as $hashtag) {
            $hName = $hashtag["text"];
            list($hStart, $hEnd) = $hashtag["indices"];
            $entities[] = array( $hStart, $hEnd, 'hashtag', '#' . $hName );
        }
        
        if (isset($r["entities"]["media"])) {
            $media = $r["entities"]["media"][0];
            $iType = $media["type"];
            if (strcasecmp($iType,"photo") == 0) {
                $iUrl = $media["url"];
                $iMUrl = $media["media_url"];
                list($iStart, $iEnd) = $media["indices"];
                $iWidth = $media["sizes"]["medium"]["w"];
                $iHeight = $media["sizes"]["medium"]["h"];
                $entities[] = array( $iStart, $iEnd, 'media', $iUrl, $iMUrl, $iWidth, $iHeight );
            }
        }

        if (isset($entities)) {
            rsort($entities);
            
            $convertedEntities = 0;
            foreach ($entities as $entity) {
                if (strcasecmp(substr($respText, $entity[0], $entity[1]-$entity[0]),$entity[3]) == 0) {
                    if (strcmp($entity[2],"url") == 0) {
                        $final = substr_replace($respText, '<a href="' . $entity[3] . '">' . $entity[3] . '</a>', $entity[0], $entity[1]-$entity[0]);
                        $convertedEntities++;
                    } elseif (strcmp($entity[2],"mention") == 0) {
                        $final = substr_replace($respText, '<a href="https://twitter.com/' . trim($entity[3],'@') . '">' . $entity[3] . '</a>', $entity[0], $entity[1]-$entity[0]);
                        $convertedEntities++;
                    } elseif (strcmp($entity[2],"hashtag") == 0) {
                        $final = substr_replace($respText, '<a href="https://twitter.com/search?q=%23' . trim($entity[3],'#') . '&amp;src=hash">' . $entity[3] . '</a>', $entity[0], $entity[1]-$entity[0]);
                        $convertedEntities++;
                    } elseif (strcmp($entity[2],"media") == 0) {
                        if (Config::EMBED_IMAGE) {
                            $final = substr_replace($respText, '<img src="' . $entity[4] . '" width="' . $entity[5] . '" height="' . $entity[6] . '" border="0">', $entity[0], $entity[1]-$entity[0]);
                        } else {
                            $final = substr_replace($respText, '<a href="' . $entity[3] . '">' . $entity[3] . '</a>', $entity[0], $entity[1]-$entity[0]);
                        }
                        $convertedEntities++;
                    }
                    $respText = $final;
                }
            }
            writeDebugMessage("INFO", "\tConverted " . $convertedEntities . " entities.");
        }

        return $respText;
    }
}

// initialize config
function initConfig() {
    if (strcmp(Config::DEBUG_MODE,"OFF") <> 0 && (strcmp(Config::DEBUG_OUTPUT,"LOG") == 0 || strcmp(Config::DEBUG_OUTPUT,"ALL") == 0)) {
        global $passedID, $passedReplies, $passedRetweets, $logHandle;
        $logfile = Config::FILE_PATH . $passedID . '.log';
        $logHandle = fopen($logfile, 'a') or die('Unable to open file: '.$logfile);
        writeDebugMessage("INFO", "Log file initialized: " . $logfile);
    }
    writeDebugMessage("INFO", "Initializing configuration...");
    $refl = new ReflectionClass('Config');
    writeDebugMessage("INFO", $refl->getConstants());
    writeDebugMessage("INFO", "Initialization complete.");
}
// clean up logging
function cleanUp() {
    if (strcmp(Config::DEBUG_MODE,"OFF") <> 0 && (strcmp(Config::DEBUG_OUTPUT,"LOG") == 0 || strcmp(Config::DEBUG_OUTPUT,"ALL") == 0)) {
        global $logHandle;
        fclose($logHandle);
        writeDebugMessage("INFO", "Log file closed.");
    }
    writeDebugMessage("INFO", "Finished.");
}

// write debug message
function writeDebugMessage($messageMode, $message) {
    if (strcmp(Config::DEBUG_MODE, $messageMode) == 0 || strcmp(Config::DEBUG_MODE, "ALL") == 0) {
        writeDebugOutput(Config::DEBUG_OUTPUT, $messageMode, $message);
    }
}
// write debug output
function writeDebugOutput($output, $messageMode, $message) {
    $ts = @date('Y-m-d H:i:s');
    global $logHandle;

    // used for config output
    if (is_array($message)) {
        $logEntry = '';
        foreach ($message as $i => $value) {
            $logEntry .= "\t[" . $i . "] => " . $value . "\n";
        }
    } else {
        $logEntry = $ts . " [" . $messageMode . "] " . $message . "\n";
    }

    switch ($output) {
        case "SCREEN":
            print $logEntry;
            break;
        case "LOG":
            fwrite($logHandle, $logEntry);
            break;
        case "ALL":
            print $logEntry;
            fwrite($logHandle, $logEntry);
            break;
    }
}

// instantiate ctor
$tmhOAuth = new tmhOAuthExample();

if (strcmp(Config::DATA_SOURCE,"LIVE") == 0) {
    writeDebugMessage("INFO", "Preparing API request.");

    if ($passedReplies == TRUE) {
        writeDebugMessage("INFO", "exclude_replies set to TRUE");
    }
    if ($passedRetweets == FALSE) {
        writeDebugMessage("INFO", "include_rts set to FALSE");
    }
    
    // call API
    $code = $tmhOAuth->request('GET', $tmhOAuth->url('1.1/statuses/user_timeline'), 
            array('screen_name' => $passedID, 'exclude_replies' => $passedReplies, 'include_rts' => $passedRetweets));
    $decoded = json_decode($tmhOAuth->response['response'], true);
    $respHeaders = $tmhOAuth->response['headers'];
    $respCode = $tmhOAuth->response['code'];
    $respInfo = $tmhOAuth->response['info'];

    /* development debug code
    $response = $tmhOAuth->response;
    print_r($response);
    echo serialize($response);
     */
} else {
    writeDebugMessage("INFO", "Preparing sample data file.");

    // no live call to API; use sample data instead; good for testing
    $testFile = Config::SAMPLE_DATA;
    $fh = fopen($testFile, "r");
    $data = unserialize(fread($fh, filesize($testFile)));
    fclose($fh);
    $decoded = json_decode($data['response'], true);
    $respHeaders = $data['headers'];
    $respCode = $data['code'];
    $respInfo = $data['info'];
}

// response code of 200 means success
// existence of "created_at" means presence of tweets
if (strcmp($respCode,'200') == 0 && isset($decoded[0]["created_at"])) {
    
    // extract twitter user data from first tweet
    $tmhOAuth->twitter_userid = $decoded[0]["user"]["screen_name"];
    $tmhOAuth->twitter_name = $decoded[0]["user"]["name"];
    $tmhOAuth->twitter_desc = $decoded[0]["user"]["description"];
    $tmhOAuth->twitter_icon = $decoded[0]["user"]["profile_image_url"];

    writeDebugMessage("INFO", 'Successfully retrieved timeline for user "' . $tmhOAuth->twitter_userid . '"');

    // start output, write header
    $tmhOAuth->openOutputFile();
    $tmhOAuth->writeHeader();
    
    writeDebugMessage("INFO", "Parsing " . count($decoded) . " tweet(s).");

    // write each tweet
    foreach ($decoded as $i) {
        writeDebugMessage("INFO", "Writing Tweet ID " . $i["id_str"] . " to RSS file...");

        // api returns the date in a format considered invalid for RSS; need to shuffle it around
        $p = explode(" ", $i["created_at"]);
        $dateFinal = $p[0] . ', ' . $p[2] . ' ' . $p[1] . ' ' . $p[5] . ' ' . $p[3] . ' ' . $p[4];

        $content = "\t\t" . '<item>' . "\n";
        if (Config::TRUNCATE_TITLE > 0) {
            $content .= "\t\t" . ' <title>' . $tmhOAuth->truncate($i["text"], Config::TRUNCATE_TITLE) . '</title>' . "\n";
        } else {
            $content .= "\t\t" . ' <title>' . $i["text"] . '</title>' . "\n";
        }
        $content .= "\t\t" . ' <link>https://twitter.com/' . $tmhOAuth->twitter_userid . '/status/' . $i["id_str"] . '</link>' . "\n";
        $content .= "\t\t" . ' <guid isPermaLink="false">https://twitter.com/' . $tmhOAuth->twitter_userid . '/status/' . $i["id_str"] . '</guid>' . "\n";
        if (Config::CONVERT_ENTITIES) {
            $content .= "\t\t" . ' <description><![CDATA[' . $tmhOAuth->links($i) . ']]></description>' . "\n";
        } else {
            $content .= "\t\t" . ' <description><![CDATA[' . $i["text"] . ']]></description>' . "\n";
        }
        $content .= "\t\t" . ' <dc:creator>' . $tmhOAuth->twitter_userid . '</dc:creator>' . "\n";
        $content .= "\t\t" . ' <pubDate>' . $dateFinal . '</pubDate>' . "\n";
        $content .= "\t\t" . '</item>' . "\n";

        $current = file_get_contents($tmhOAuth->file);
        $current .= $content;
        file_put_contents($tmhOAuth->file, $current);
    }

    $tmhOAuth->writeFooter();

} elseif (strcmp($respCode,'200') == 0 && count($decoded) == 0) {
    // a rare case, but if a valid user hasn't tweeted yet, API will still return a 200
    writeDebugMessage("INFO", 'User "' . $passedID . '" has not tweeted anything yet!');
} else {
    // errors
    writeDebugMessage("INFO", 'User timeline retrieval unsuccessful for user "' . $passedID . '". Set DEBUG_MODE: ERROR or ALL for more info.');
    if (isset($decoded["errors"][0]["message"])) {
        $allKeys = array_keys($respHeaders);
        writeDebugMessage("ERROR", $allKeys[0] . ' "' . $decoded["errors"][0]["message"] . '"');
    }
}

// adios
cleanUp();

?>
